loader/loader.go

```diff
@@ -1188,110 +1188,246 @@ func fetchMatchedLiteral(ctx *tcontext.Context, router *router.Table, schema, ta
 
-		tnames := make([]string, 0, len(tables))
-		for t := range tables {
-			tnames = append(tnames, t)
-		}
-		for _, table := range tnames {
-			dataFiles := tables[table]
-			tableFile := fmt.Sprintf("%s/%s.%s-schema.sql", l.cfg.Dir, db, table)
+	// push table schema restoring jobs to the queue
+tblSchemaLoop:
+	for dbSessionID, db := range dbs {
+		dbSessionID %= concurrency
+		for table := range l.db2Tables[db] {
+			schemaFile := l.cfg.Dir + "/" + db + "." + table + "-schema.sql" // cache friendly
 			if _, ok := l.tableInfos[tableName(db, table)]; !ok {
-				l.tableInfos[tableName(db, table)], err = parseTable(tctx, l.tableRouter, db, table, tableFile, l.cfg.LoaderConfig.SQLMode)
+				l.tableInfos[tableName(db, table)], err = parseTable(tctx, l.tableRouter, db, table, schemaFile, l.cfg.LoaderConfig.SQLMode)
 				if err != nil {
-					return terror.Annotatef(err, "parse table %s/%s", db, table)
+					err = terror.Annotatef(err, "parse table %s/%s", db, table)
+					break tblSchemaLoop
 				}
 			}
-
 			if l.checkPoint.IsTableFinished(db, table) {
 				l.logger.Info("table has finished, skip it.", zap.String("schema", db), zap.String("table", table))
 				continue
 			}
-
-			// create table
-			l.logger.Info("start to create table", zap.String("table file", tableFile))
-			err := l.restoreTable(ctx, dbConn, tableFile, db, table)
+			err = tblRestoreQueue.push(&restoreSchemaJob{
+				loader:   l,
+				session:  dbSessionPool[dbSessionID],
+				database: db,
+				table:    table,
+				filepath: schemaFile,
+			})
 			if err != nil {
-				return err
+				break tblSchemaLoop
 			}
-			l.logger.Info("finish to create table", zap.String("table file", tableFile))
+		}
+	}
 
-			restoringFiles := l.checkPoint.GetRestoringFileInfo(db, table)
-			l.logger.Debug("restoring table data", zap.String("schema", db), zap.String("table", table), zap.Reflect("data files", restoringFiles))
+	// check producing error
+	if err != nil {
+		runtimeErr := tblRestoreQueue.close()
+		if errors.ErrorEqual(err, context.Canceled) {
+			err = runtimeErr
+		}
+		return err
+	}
+	// wait whole task done & close queue
+	err = tblRestoreQueue.close()
+	if err != nil {
+		return err
+	}
+
+	// all schemas was restored
+	l.logger.Info("finish to create tables", zap.Duration("cost time", time.Since(begin)))
 
-			info := l.tableInfos[tableName(db, table)]
-			for _, file := range dataFiles {
+	// generate restore table data file job
+	for _, db := range dbs {
+		table2DataFileMap := l.db2Tables[db]
+		for table := range table2DataFileMap {
+			for _, file := range table2DataFileMap[table] {
 				select {
 				case <-ctx.Done():
 					l.logger.Warn("stop generate data file job", log.ShortError(ctx.Err()))
 					return ctx.Err()
 				default:
 					// do nothing
 				}
-
+				restoringFiles := l.checkPoint.GetRestoringFileInfo(db, table)
+				l.logger.Debug("restoring table data", zap.String("schema", db), zap.String("table", table), zap.Reflect("data files", restoringFiles))
```

Comment:
Suggest putting these two lines between these lines. Is `restoringFiles` any difference for each file?
```
		for table := range table2DataFileMap {
			for _, file := range table2DataFileMap[table] { 
```

Reply:
Well it's retrieved from checkpoint, I'm not so sure. BTW, it's same logic as before. would you like take a look? @lance6716

Reply:
```
restoringFiles := l.checkPoint.GetRestoringFileInfo(db, table)
```

doesn't involve `file`, so it could be moved out of `file` loop

Reply:
> ```
> restoringFiles := l.checkPoint.GetRestoringFileInfo(db, table)
> ```
> 
> doesn't involve `file`, so it could be moved out of `file` loop

nice